src/Yaapii.Xambly/Directive/NsDirective.cs

```diff
@@ -77,45 +172,169 @@ public override string ToString()
         /// <param name="stack">Execution stack</param>
         /// <returns>New current nodes</returns>
         public ICursor Exec(XNode dom, ICursor cursor, IStack stack)
-        {
-            throw new ImpossibleModificationException("Modifying namespaces is not supported at the moment.");
+        {
             try
             {
-                XElement element = null;
-                if (dom is XDocument)
-                {
-                    element = (dom as XDocument).Root;
-                }
-                else
-                {
-                    element = dom as XElement;
+                if (this.prefix.Value() == string.Empty)
+                {
+                    this.ApplyDefaultNS(dom, cursor, stack);
                 }
-
-                new FailPrecise(
-                    new FailNull(element),
-                    new ArgumentException($"Node is not of type 'XElement'")
-                ).Go();
-
-                XNamespace xnsp = this.nsp.Raw();
-
-                ApplyNamespace(element, xnsp);
-
+                else
+                {
+                    this.ApplyPrefixedNS(dom, cursor, stack);
+                }
                 return cursor;
             }
             catch (XmlContentException ex)
             {
-                throw new IllegalArgumentException("can't set xmlns", ex);
+                throw new IllegalArgumentException($"Failed to understand XML content, {this}", ex);
```

Comment:
The error is, that the namespace cannot be applied. The content of the exception is the reason. 
So please change back to the previous expression. Currently, the Exception is misleading.